src/test/java/io/jenkins/plugins/credentials/secretsmanager/GithubAppCredentialsIT.java

```diff
@@ -0,0 +1,213 @@
+
+        final GitHubAppCredentials theirs = new GitHubAppCredentials(CredentialsScope.GLOBAL, "name", "description", "11111", Secret.fromString("secret"));
+
+        assertThat(ours)
+                .hasSameDescriptorIconAs(theirs);
+    }
+
+    @Test
+    @ConfiguredWithCode(value = "/integration.yml")
+    public void shouldSupportWithCredentialsBinding() {
+        // Given
+        final CreateSecretResult foo = createGitHubAppCredentialSecret(APP_ID, PRIVATE_KEY);
+
+        new MockUp<GitHubAppCredentials>() {
+            @Mock
+            public Secret getPassword() {
+                return Secret.fromString("test-token");
+            }
+        };
+
+        // When
+        final WorkflowRun run = runPipeline("",
+                "withCredentials([usernamePassword(credentialsId: '" + foo.getName() + "', usernameVariable: 'USR', passwordVariable: 'PSW')]) {",
+                "  echo \"Credential: {username: $USR, password: $PSW}\"",
+                "}");
+
+        // Then
+        assertThat(run)
+                .hasResult(hudson.model.Result.SUCCESS)
+                .hasLogContaining("Credential: {username: ****, password: ****}");
+    }
+
+    @Test
+    @ConfiguredWithCode(value = "/integration.yml")
+    public void shouldSupportEnvironmentBinding(){
+        // Given
+        final CreateSecretResult foo = createGitHubAppCredentialSecret(APP_ID, PRIVATE_KEY);
+
+        new MockUp<GitHubAppCredentials>() {
+            @Mock
+            public Secret getPassword() {
+                return Secret.fromString("test-token");
+            }
+        };
+
+        // When
+        final WorkflowRun run = runPipeline("",
+                "pipeline {",
+                "  agent none",
+                "  stages {",
+                "    stage('Example') {",
+                "      environment {",
+                "        FOO = credentials('" + foo.getName() + "')",
+                "      }",
+                "      steps {",
+                "        echo \"{variable: $FOO, username: $FOO_USR, password: $FOO_PSW}\"",
+                "      }",
+                "    }",
+                "  }",
+                "}");
+
+        // Then
+        assertThat(run)
+                .hasResult(hudson.model.Result.SUCCESS)
+                .hasLogContaining("{variable: ****, username: ****, password: ****}");
+    }
+
+    @Test
+    @ConfiguredWithCode(value = "/integration.yml")
+    public void shouldSupportSnapshots() {
+        // Given
+        final CreateSecretResult foo = createGitHubAppCredentialSecret(APP_ID, PRIVATE_KEY);
+        final GitHubAppCredentials before = jenkins.getCredentials().lookup(GitHubAppCredentials.class, foo.getName());
+
+        // When
+        final GitHubAppCredentials after = CredentialSnapshots.snapshot(before);
+
+        // Then
+        assertThat(after)
+                .hasUsername(before.getUsername())
+                .hasId(before.getId());

```

Comment:
Can you test the retrieval of the password / secret field? We need this to ensure that snapshotting of the secret value works correctly, which is necessary for distributed Jenkins setups. You might need to write a custom AssertJ matcher, see these classes for examples:

- https://github.com/jenkinsci/aws-secrets-manager-credentials-provider-plugin/blob/2229d682e4135a844147f88316a6c675b37f8b5f/src/test/java/io/jenkins/plugins/credentials/secretsmanager/SSHUserPrivateKeyIT.java#L182 (a test that uses the matcher)
- https://github.com/jenkinsci/aws-secrets-manager-credentials-provider-plugin/blob/2229d682e4135a844147f88316a6c675b37f8b5f/src/test/java/io/jenkins/plugins/credentials/secretsmanager/util/assertions/SSHUserPrivateKeyAssert.java#L11 (the custom matcher)

Reply:
Ok, I added a new Assert for GithubAppCredentials, and we can test the private key with the snapshot now.

Reply:
Is there anything else that needs to be added?